rollbar/lib/_async.py

```diff
@@ -0,0 +1,268 @@
+    async with httpx.AsyncClient(
+        proxies=proxies, verify=rollbar.SETTINGS.get('verify_https', True)
+    ) as client:
+        resp = await client.post(
+            url,
+            data=payload_str,
+            headers=headers,
+            timeout=rollbar.SETTINGS.get('timeout', DEFAULT_TIMEOUT),
+        )
+
+    try:
+        return rollbar._parse_response(path, access_token, payload_str, resp)
+    except Exception as e:
+        log.exception('Exception while posting item %r', e)
+
+
+async def try_report(
+    exc_info=None, request=None, extra_data=None, payload_data=None, level=None, **kw
+):
+    current_handler = rollbar.SETTINGS.get('handler')
+    if not (current_handler in ALLOWED_HANDLERS or current_handler == 'default'):
+        raise RollbarAsyncError('No async handler set.')
+
+    if httpx is None:
+        raise RollbarAsyncError('HTTPX is required')
+
+    return await report_exc_info(
+        exc_info, request, extra_data, payload_data, level, **kw
+    )
+
+
+class AsyncHandler:
+    def __init__(self):
+        self.global_handler = None
+        self.token = None
+
+    def with_ctx_handler(self):
+        if self.global_handler in ALLOWED_HANDLERS:
+            self.token = _ctx_handler.set(self.global_handler)
+        else:
+            log.warning(
+                'Running coroutines requires async compatible handler. Switching to default async handler.'
+            )
+            self.token = _ctx_handler.set('async')
+
+        return _ctx_handler.get()
+
+    def with_global_handler(self):
+        return self.global_handler
+
+    def __enter__(self):
+        self.global_handler = rollbar.SETTINGS.get('handler')
+
+        if _ctx_handler:
+            return self.with_ctx_handler()
+        else:
+            return self.with_global_handler()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if _ctx_handler and self.token:
+            _ctx_handler.reset(self.token)
+
+
+def get_current_handler():
+    if _ctx_handler is None:
+        return rollbar.SETTINGS.get('handler')
+
+    handler = _ctx_handler.get()
+
+    if handler is None:
+        return rollbar.SETTINGS.get('handler')
+
+    return handler
+
+
+def call_later(coro):
+    if sys.version_info < (3, 7):
+        return asyncio.ensure_future(coro)
+
+    return asyncio.create_task(coro)
+
+
+# test helpers
+# TODO: move to rollbar.test.async_helper after migrating from unittest
+
+
+def run(coro):
+    if sys.version_info >= (3, 7):
+        return asyncio.run(coro)
+
+    assert inspect.iscoroutine(coro)
+
+    loop = asyncio.new_event_loop()
+    asyncio.set_event_loop(loop)
+    try:
+        return loop.run_until_complete(coro)
+    finally:
+        loop.close()
+        asyncio.set_event_loop(None)
+
+
+def async_receive(message):
+    async def receive():
+        return message
+
+    assert message['type'] == 'http.request'
+    return receive
+
+
+async def coroutine():
```

Comment:
This seems to only be used in tests, should it be moved to a test module?

Reply:
You are right but there is nothing we can do about it at this time. The current test runner is not able to handle it properly. Here is the commit: 1507489